### PR TITLE
[MAINTENANCE] Standardize `Datasource` CRUD

### PR DESCRIPTION
**Pull request by cdkini**
Changes proposed in this pull request:
- Add `add_or_update_datasource`

### Definition of Done
- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/docs/contributing/contributing_checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added [unit tests](https://docs.greatexpectations.io/docs/contributing/contributing_test#writing-unit-and-integration-tests) where applicable and made sure that new and existing tests are passing.
- [x] I have run any local integration tests and made sure that nothing is broken.

### CODE REVIEW

File: great_expectations/data_context/data_context/abstract_data_context.py [hunk 1]
```diff
@@ -730,17 +730,60 @@ def update_datasource(
...
+        )
+        return datasource
+
+    def add_or_update_datasource(
+        self,
+        name: str,
+        **kwargs,
+    ) -> LegacyDatasource | BaseDatasource:
+        """
+        TODO
```
Comment (cdkini) [inline great_expectations/data_context/data_context/abstract_data_context.py:762]: Open to implement

File: great_expectations/data_context/data_context/abstract_data_context.py [hunk 2]
```diff
@@ -730,17 +730,70 @@ def update_datasource(
...
         if save_changes:
             self._datasource_store.update_by_name(  # type: ignore[attr-defined]
-                datasource_name=datasource_name, datasource_config=datasource_config
+                datasource_name=name, datasource_config=datasource_config
+            )
+
+        datasource = self._instantiate_datasource_from_config_and_update_project_config(
+            config=datasource_config, initialize=True, save_changes=False
+        )
+        assert datasource is not None
```
Comment (cdkini) [inline great_expectations/data_context/data_context/abstract_data_context.py:754]: Invariant based on `initialize=True` above

File: great_expectations/data_context/data_context/abstract_data_context.py [hunk 3]
```diff
@@ -735,27 +735,80 @@ def add_datasource(
...
-            datasource_config: The config object to persist using the DatasourceStore.
+            datasource: The Datasource object to update.
             save_changes: do I save changes to disk?
         """
         save_changes = self._determine_save_changes_flag(save_changes)
+        self._update_datasource(datasource=datasource, save_changes=save_changes)
+
+    def _update_datasource(
+        self,
+        datasource: Union[LegacyDatasource, BaseDatasource],
```
Comment (Kilo59) [inline great_expectations/data_context/data_context/abstract_data_context.py]:
```suggestion
        datasource: LegacyDatasource | BaseDatasource,
```
non-blocking

File: great_expectations/data_context/data_context/abstract_data_context.py [hunk 4]
```diff
@@ -735,27 +735,80 @@ def add_datasource(
...
+        save_changes: bool | None = None,
+        **kwargs,
+    ) -> Datasource:
+        name = datasource.name
+        config = datasource.config
+        # `instantiate_class_from_config` requires `class_name`
+        config["class_name"] = datasource.__class__.__name__
+        config.update(kwargs)
```
Comment (Kilo59) [inline great_expectations/data_context/data_context/abstract_data_context.py:787]: `non-blocking`
Would there be any reason to raise an error if no `kwargs` are passed.

**Review by Kilo59 — APPROVED** [review]
non-blocking questions/comments

**Pull request merged**